tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -9720,6 +9720,69 @@ void hlsl::DiagnosePackingOffset(
   }
 }
 
+static bool IsIntegerTypeCSInputSemantic(const hlsl::SemanticDecl *sd) {
+  return
+    sd->SemanticName.equals("SV_DispatchThreadID") ||
+    sd->SemanticName.equals("SV_GroupIndex") ||
+    sd->SemanticName.equals("SV_GroupID") ||
+    sd->SemanticName.equals("SV_GroupThreadID");
+}
+
+static bool IsComputeShader(clang::Sema *self) {
+  StringRef shaderType(self->getLangOpts().HLSLProfile);
+  return shaderType.startswith_lower("cs_");
+}
+
+static bool
+HasValidCSInputSemanticType(clang::Sema *self, QualType &ty,
+                            const hlsl::SemanticDecl *semanticDecl) {
+  if (!IsIntegerTypeCSInputSemantic(semanticDecl))
+    return true;
+
+  QualType qty = ty.getCanonicalType();
+
+  const clang::Type *qtyPtr = qty.getTypePtrOrNull();
+  if (qtyPtr && qtyPtr->isScalarType() &&
+    qtyPtr->getScalarTypeKind() != Type::ScalarTypeKind::STK_Integral) {
+    return false;
+  }
```

Comment:
The code might be more self-documenting and less repetitive if this common check were parceled off into a small subfunction.